rust/src/ikev1/parser.rs

```diff
@@ -0,0 +1,732 @@
+            _ => write!(f, "unknown")
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum AttributeValue {
+    // https://www.iana.org/assignments/ipsec-registry/ipsec-registry.xhtml
+    Unknown,
+    // Encryption Algorithm
+    EaDesCbc,
+    EaIdeaCbc,
+    EaBlowfishCbc,
+    EaRc5R16B64Cbc,
+    EaTripleDesCbc,
+    EaCastCbc,
+    EaAesCbc,
+    EaCamelliaCbc,
+    // Hash Algorithm
+    HashMd5,
+    HashSha,
+    HashTiger,
+    HashSha2_256,
+    HashSha2_384,
+    HashSha2_512,
+    // Authentication Method
+    AuthPreSharedKey,
+    AuthDssSignatures,
+    AuthRsaSignatures,
+    AuthEncryptionWithRsa,
+    AuthRevisedEncryptionWithRsa,
+    AuthReserved,
+    AuthEcdsaSha256,
+    AuthEcdsaSha384,
+    AuthEcdsaSha512,
+    // Group Description
+    GroupDefault768BitModp,
+    GroupAlternate1024BitModpGroup,
+    GroupEc2nOnGp2p155,
+    GroupEc2nOnGp2p185,
+    GroupModp1536Bit,
+    GroupEc2nOverGf2p163,
+    GroupEc2nOverGf2p283,
+    GroupEc2nOverGf2p409,
+    GroupEc2nOverGf2p571,
+    GroupModp2048Bit,
+    GroupModp3072Bit,
+    GroupModp4096Bit,
+    GroupModp6144Bit,
+    GroupModp8192Bit,
+    GroupRandomEcp256,
+    GroupRandomEcp384,
+    GroupRandomEcp521,
+    GroupModp1024With160BitPrime,
+    GroupModp2048With224BitPrime,
+    GroupModp2048With256BitPrime,
+    GroupRandomEcp192,
+    GroupRandomEcp224,
+    GroupBrainpoolEcp224,
+    GroupBrainpoolEcp256,
+    GroupBrainpoolEcp384,
+    GroupBrainpoolEcp512,
+    // Life Type
+    LifeTypeSeconds,
+    LifeTypeKilobytes
+}
+
+impl fmt::Display for AttributeValue {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+            write!(f, "{:?}", self)
+        }
+}
+
+pub struct SaAttribute {
+    pub attribute_format: u8,
+    pub attribute_type: AttributeType,
+    pub attribute_value: AttributeValue,
+    pub numeric_value: Option<u32>,
+    pub hex_value: Option<String>
+}
+
+named! {pub parse_isakmp_header<IsakmpHeader>,
+    do_parse!(
+        init_spi: be_u64 >>
+        resp_spi: be_u64 >>
+        np:       be_u8 >>
+        vers: bits!(
+            tuple!(take_bits!(4u8),take_bits!(4u8))
+        ) >>
+        ex:       be_u8 >>
+        flags:    be_u8 >>
+        id:       be_u32 >>
+        l:        be_u32 >>
+        (
+            IsakmpHeader {
+                init_spi,
+                resp_spi,
+                next_payload: np,
+                maj_ver: vers.0,
+                min_ver: vers.1,
+                exch_type: ex,
+                flags,
+                msg_id: id,
+                length: l,
+            }
+        )
+    )
+}
+
+pub fn parse_security_association(
+    i: &[u8],
+    length: u16,
+) -> IResult<&[u8], SecurityAssociationPayload> {
+    do_parse!(
+        i,
+        domain_of_interpretation: be_u32 >>
+        situation: cond!(domain_of_interpretation == 1, take!(4)) >>
+        data: cond!(domain_of_interpretation == 1, take!(length - 8)) >>
```

Comment:
You must test `length >= 8` before doing that, or you may cause an integer underflow here.